src/analysis/retail/warrior/arms/modules/features/RageTracker.tsx

```diff
@@ -0,0 +1,85 @@
+import SPELLS from 'common/SPELLS';
+import TALENTS from 'common/TALENTS/warrior';
+import HIT_TYPES from 'game/HIT_TYPES';
+import RESOURCE_TYPES from 'game/RESOURCE_TYPES';
+import { Options, SELECTED_PLAYER } from 'parser/core/Analyzer';
+import Events, { CastEvent, DamageEvent } from 'parser/core/Events';
+import ResourceTracker from 'parser/shared/modules/resources/resourcetracker/ResourceTracker';
+
+const RAGE_PER_MELEE_HIT = 25;
+
+class RageUsage extends ResourceTracker {
+  lastMeleeTimestamp = 0;
+
+  private usesWarlordsTorment = false;
+  private usesWarMachine = false;
+
+  constructor(options: Options) {
+    super(options);
+    this.resource = RESOURCE_TYPES.RAGE;
+    this.usesWarlordsTorment = this.selectedCombatant.hasTalent(TALENTS.WARLORDS_TORMENT_TALENT);
+    this.usesWarMachine = this.selectedCombatant.hasTalent(TALENTS.WAR_MACHINE_ARMS_TALENT);
+    this.addEventListener(Events.damage.by(SELECTED_PLAYER).spell(SPELLS.MELEE), this.onDamage);
+    //the rage refund from improved execute.
+    if (!this.selectedCombatant.hasTalent(TALENTS.IMPROVED_EXECUTE_ARMS_TALENT)) {
+      return;
+    }
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(SPELLS.EXECUTE_GLYPHED),
+      this.onCastExecute,
+    );
+  }
+
+  getAdjustedCost(event: CastEvent) {
+    if (event.resourceCost && event.resourceCost[this.resource.id] !== undefined) {
+      return event.resourceCost[this.resource.id];
+    }
+    const resource = super.getResource(event);
+    if (!resource) {
+      return;
+    }
+    return resource.cost;
+  }
+
+  onCastExecute(event: CastEvent) {
+    //the cost for execute is multiplied by 10 to avoid floating point numbers.
+    //just divide it by 10 again to get the more accurate number.
+    const cost = (this.getAdjustedCost(event) || 0) / 10;
+    if (cost !== 0) {
+      this.processInvisibleEnergize(
+        SPELLS.EXECUTE.id,
+        cost / 10, //always assume the target survives, get 10% refund.
+        event.timestamp,
+      );
+    }
+  }
```

Comment:
don't need this as part of this PR because this will be *very* close to correct, but you can use a `CastLinkNormalizer` to connect the execute cast to the execute damage event, then check if the execute damage event had `overkill > 0` to see if you should refund.

Reply:
Thats a good suggestion, thanks!
But I think I would like this to stay here until I get around to adding a CastLinkNormalizer to do just that. The little bit of extra rage tracking accuracy feels pretty good to have in my opinion.

BTW the merge is on its way

Reply:
just to confirm: "I would like this to stay here" means don't merge til you get the normalizer work done? or merge now and normalizer work later?

Reply:
Oh, it couldve gotten merged already, but i'll try to get around to getting the normalizer done until monday if i got time, so either i add it to this, or ill open a seperate PR for the normalizer.